axelor-stock/src/main/java/com/axelor/apps/stock/service/StockLocationServiceImpl.java

```diff
@@ -90,89 +103,84 @@ protected List<StockLocation> getNonVirtualStockLocations(Long companyId) {
+  public BigDecimal getQtyOfProductInStockLocations(
+      Long productId, List<Long> stockLocationIds, Long companyId, String qtyFieldName)
       throws AxelorException {
-    if (productId == null) {
-      return BigDecimal.ZERO;
-    }
-
     Product product = productRepo.find(productId);
     Unit productUnit = product.getUnit();
-    UnitConversionService unitConversionService = Beans.get(UnitConversionService.class);
-    int scale = Beans.get(AppBaseService.class).getNbDecimalDigitForQty();
-    BigDecimal qty = BigDecimal.ZERO;
-
-    if (locationId == null || locationId == 0L) {
-      List<StockLocation> stockLocations = getNonVirtualStockLocations(companyId);
-      if (stockLocations.isEmpty()) {
-        return BigDecimal.ZERO;
-      }
-
-      for (StockLocation stockLocation : stockLocations) {
-        StockLocationLine stockLocationLine =
-            stockLocationLineService.getStockLocationLine(
-                stockLocationRepo.find(stockLocation.getId()), productRepo.find(productId));
-        if (stockLocationLine == null) {
-          continue;
-        }
-
-        Unit stockLocationLineUnit = stockLocationLine.getUnit();
-        qty =
-            qty.add(
-                qtyType.equals("real")
-                    ? stockLocationLine.getCurrentQty()
-                    : stockLocationLine.getFutureQty());
-
-        if (productUnit != null && !productUnit.equals(stockLocationLineUnit)) {
-          qty =
-              unitConversionService.convert(
-                  stockLocationLineUnit, productUnit, qty, qty.scale(), product);
-        }
-      }
-
-    } else {
-      StockLocationLine stockLocationLine =
-          stockLocationLineService.getStockLocationLine(
-              stockLocationRepo.find(locationId), productRepo.find(productId));
 
-      if (stockLocationLine == null) {
-        return BigDecimal.ZERO;
-      }
+    int scale = appBaseService.getNbDecimalDigitForQty();
 
-      Unit stockLocationLineUnit = stockLocationLine.getUnit();
-      qty =
-          qtyType.equals("real")
-              ? stockLocationLine.getCurrentQty()
-              : stockLocationLine.getFutureQty();
+    String query =
+        "SELECT sll.unit.id, sum(sll.%1$s)"
+            + " FROM StockLocationLine sll"
+            + " WHERE (( (0 IN (%2$s)) AND sll.stockLocation.typeSelect != :stockLocationTypeSelectVirtual AND (sll.stockLocation.company.id = :companyId OR COALESCE(:companyId,0) = 0 ) )"
+            + " OR (sll.stockLocation.id in (%2$s) AND 0 NOT IN (%2$s) ))"
+            + " AND :productId is not null AND sll.product.id = :productId AND sll.product.stockManaged is TRUE GROUP BY sll.unit.id";
 
-      if (productUnit != null && !productUnit.equals(stockLocationLineUnit)) {
-        qty =
-            unitConversionService.convert(
-                stockLocationLineUnit, productUnit, qty, qty.scale(), product);
+    List<Tuple> sumOfQtyPerUnitList =
```

Comment:
I don't get why it is a List<Tuple> and not a Map<Long, BigDecimal>, can you check ?

Reply:
`getResultList` returns a list of a generic type of our choice. so it's natural to use a list here.
What do you suggest i do otherwise ?

Reply:
I thought a map could be used, but after checking existing code, I did not find something better (we can do a List of Map, but it would not simplify the code)